feature: new tests added for tsne to expand test coverage

yuejiaointel opened the pull request:
## Description

Added additional tests in sklearnex/manifold/tests/test_tsne.py to expand the test coverage for t-SNE algorithm. The associated issue number is 8325, https://jira.devtools.intel.com/browse/DAALL-8325


**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [ ] All CI jobs are green or I have provided justification why they aren't.
- [ ] I have extended testing suite if new functionality was introduced in this PR.

icfaust commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -16,11 +16,175 @@
 from numpy.testing import assert_allclose
-
+import pytest
+#Note: n_componets must be 2 for now
```
```suggestion
#Note: n_components must be 2 for now

```

yuejiaointel commented:
/intelci: run

ethanglaser commented:
/intelci: run

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+    tsne_perplexity = TSNE(n_components=2, perplexity=9).fit(X_perplexity)
+    assert tsne_perplexity.embedding_.shape == (10, 2)
+
+    # Test large data
```
It feels like this one is perhaps not needed, considering that there's already a similar test earlier on with shape (100,10).

yuejiaointel replied:
Hi David, I removed this test. Best, Yue

david-cortes-intel replied:
Thanks, although if the "Test reproducibility" test is also checking the shape, it still feels like "Test large data" could be removed altogether since it's not testing anything different. Or does the algorithm have some size-dependent behavior that would change between `(50,10)` and (1000, 50)`?

yuejiaointel replied:
Hi David, 
Thx for the suggestion! I removed test large data, it used to be much larger but I agree 1000 and 50 are not that different
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+    X_constant = np.ones((10, 10))
+    tsne = TSNE(n_components=2, perplexity=5, random_state=42)
+    embedding = tsne.fit(X_constant).embedding_
+    assert embedding.shape == (10, 2), f"Unexpected embedding shape: {embedding.shape}"
```
I believe `pytest` already does the substitution for you in the error message when the test fails.

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+
+def test_basic_tsne_functionality():
+    """Test TSNE with valid data: basic functionality, random data, reproducibility, and edge cases."""
+    # Test basic functionality
```
Any chance that all of these could be parameterized instead? Otherwise, if one of them fails, then the rest wouldn't execute.

david-cortes-intel replied:
Thanks for looking into it. But what I meant was to parameterize this whole function by turning the inputs and expectations, so that one parameterization would be "Test basic functionality", another parameterization "Test with random data", and so on.

yuejiaointel replied:
Hi David,
Thx for the suggestion, I refactored this function again to have parametrize each sub test, hopefully this is the right idea  
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+
+    # Edge Case: Sparse-Like High-Dimensional Data
+    np.random.seed(42)
+    X_sparse_like = np.random.rand(50, 10000) * (np.random.rand(50, 10000) > 0.99)
```
Perhaps 10k is a bit too large for _just_ a test case. Could it be shrunk to e.g. 500 or so? If needed, perhaps you could use the random generator from `scipy.sparse` and then cast it to dense.

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_tsne_gpu_validation(dataframe, queue, dtype):
+    """
+    GPU validation test for TSNE with a specific complex dataset.
```
Could this one be merged into the one above? (`test_tsne_with_specific_complex_dataset`)

david-cortes-intel replied:
Looks like the tests for non-missingness were lost after merging them. Could be good to have them just in case, considering that it tests GPU execution where more things can go wrong.

yuejiaointel replied:
Hi David, 
Thx for the comments, I brought back the deleted test and merged with the previous complex test, one of the check is only valid on gpu device so I moved get_dataframes_and_queues from decorator to code to accept filter device as a parameter.
Best, Yue

yuejiaointel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+import pytest
 from numpy.testing import assert_allclose
 
+# Note: n_components must be 2 for now
```
Hi David, 
Thx for the comments, I made a new commit that addressed other comments but for this one, Ethan double checked with Victoriya that TSNE works only for n_components == 2, here is the specific message:
yes our TSNE works only for n_components == 2.
See in DAAL (though the error message is not very descriptive): https://github.com/uxlfoundation/oneDAL/blob/main/cpp/daal/src/algorithms/tsne/tsne_gradient_descent_impl.i#L1000
 
The reason of this behavior is that TSNE is a visualization algorithm. Its results are plotted to see the hidden structures in originally multidimensional data (usually 2d plots are used, more rarely - 3d).
So, having TSNE patched only for n_components == 2 looks like a reasonable compromise.

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,192 @@
+    assert tsne.embedding_.shape == (4, 2)
+
+    # Test with random data
+    np.random.seed(42)
```
What I meant here was this module: https://numpy.org/doc/stable/reference/random/generator.html

To use it, you need to generate a `Generator` object, for example like this:
```python
rng = np.random.default_rng(seed=123)
```

Then, you need to call methods from that generator object, like this:
```python
rng.random(size=(100,10))
```

yuejiaointel replied:
Hi David, 
Thx for the suggestion, I fixed the code to use rng generator for all sub tests in test_tsne_functionality_and_edge_cases for independent results.
Best,
Yue

yuejiaointel commented:
/intelci: run

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,242 @@
+            False,
+        ),
+        (
+            "Test reproducibility",
```
I think the reproducibility test here was lost.

yuejiaointel replied:
Hi David!
Good catch I moved reproducibility test to a new test function test_tsne_reproducibility to avoid ad hoc code in test_tsne_functionality_and_edge_cases
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,242 @@
+            assert (
+                embedding.shape == expected_shape
+            ), f"{description}: Incorrect embedding shape."
+            if device_filter == "gpu":
```
I think this doesn't need to be specific to GPU.

yuejiaointel replied:
Hi David,
You are right, I tried on my testing machine again and somehow it works this time, must be some env settings not correctly yesterday. I removed the device filter and put the parametrization decorator back for get_dataframes_and_queues
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,242 @@
+        ),
+        (
+            "Edge Case: Sparse-Like High-Dimensional Data",
+            lambda rng: rng.random((50, 500)) * (rng.random((50, 500)) > 0.99),
```
For less-random reproducibility, perhaps it could add a test where one column is constant or an exact duplicate of another column.

david-cortes-intel commented:
It looks like we don't have any test here nor in daal4py that would be checking that the results from TSNE make sense beyond having the right shape and non-missingness.

Since there's a very particular dataset here for the last test, it'd be helpful to add other assertions there along the lines of checking that the embeddings end up making some points closer than others as would be expected given the input data.

yuejiaointel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,242 @@
+    "description,X_generator,n_components,perplexity,expected_shape,should_raise",
+    [
+        (
+            "Test basic functionality",
```
Hi David,
Thx for the pointer! I read the doc and used pytest.param and put the names as id. Hope this is the right idea!
Best,
Yue

yuejiaointel commented:
Hi David, 
About the last comment, I think that is a good test to add! I spent some time thinking it through and have added a logic check in the final test to evaluate the overlap of close neighbors. Here’s a summary of the steps I implemented:
1. get a distance array where [i, j] is Euclidean distance of point i and j in original space, same for tsne embedding space
2. rank distances for each point wrt first column in original space, also for embedding space
3. get top 5 neighbors of  each point in original and embedding space see how many are same by dividing them
4. get a mean of all fractions it should represent how the original and embedding space are similar for the most 5 closest points
5. check if that mean is > 0.6
Let me know your thoughts on this approach or if you believe it could be improved further.
Thx a lot :D
Yue

david-cortes-intel commented:
> Hi David, About the last comment, I think that is a good test to add! I spent some time thinking it through and have added a logic check in the final test to evaluate the overlap of close neighbors. Here’s a summary of the steps I implemented:
> 
> 1. get a distance array where [i, j] is Euclidean distance of point i and j in original space, same for tsne embedding space
> 2. rank distances for each point wrt first column in original space, also for embedding space
> 3. get top 5 neighbors of  each point in original and embedding space see how many are same by dividing them
> 4. get a mean of all fractions it should represent how the original and embedding space are similar for the most 5 closest points
> 5. check if that mean is > 0.6
>    Let me know your thoughts on this approach or if you believe it could be improved further.
>    Thx a lot :D
>    Yue

I think given the characteristics of the data that you are passing, it could be done by selecting some hard-coded set of points by index from "Complex Dataset1" that should end up being similar, and some selected set of points that should end up being dissimilar to the earlier ones; with the test then checking that the euclidean distances in the embedding space among each point from the first set are smaller than the distances between each point in the first set and each point in the second set.

Also maybe "Complex Dataset2" is not needed.

yuejiaointel commented:
> > Hi David, About the last comment, I think that is a good test to add! I spent some time thinking it through and have added a logic check in the final test to evaluate the overlap of close neighbors. Here’s a summary of the steps I implemented:
> > 
> > 1. get a distance array where [i, j] is Euclidean distance of point i and j in original space, same for tsne embedding space
> > 2. rank distances for each point wrt first column in original space, also for embedding space
> > 3. get top 5 neighbors of  each point in original and embedding space see how many are same by dividing them
> > 4. get a mean of all fractions it should represent how the original and embedding space are similar for the most 5 closest points
> > 5. check if that mean is > 0.6
> >    Let me know your thoughts on this approach or if you believe it could be improved further.
> >    Thx a lot :D
> >    Yue
> 
> I think given the characteristics of the data that you are passing, it could be done by selecting some hard-coded set of points by index from "Complex Dataset1" that should end up being similar, and some selected set of points that should end up being dissimilar to the earlier ones; with the test then checking that the euclidean distances in the embedding space among each point from the first set are smaller than the distances between each point in the first set and each point in the second set.
> 
> Also maybe "Complex Dataset2" is not needed.

Hi David!
I fixed the logic based on your suggestion. Now it will try to get a group A with similar points and group B with different points from group A, then I check in embedding space distance bt any 2 points in group A should be less than that point to any point in group B. I hope I understand you correctly. I also removed complex test 2.
Best,
Yue

yuejiaointel commented:
/intelci: run

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+    assert_allclose(tsne_1, tsne_2, rtol=1e-5)
+
+
+def compute_pairwise_distances(data):
```
Better to use the implementation in this same package:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/metrics/pairwise.py

Or failing that, in the dependencies, like these:
https://scikit-learn.org/dev/modules/generated/sklearn.metrics.pairwise_distances.html
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.euclidean.html

yuejiaointel replied:
Hi David,
That is good note! I used the sklearn one and it works
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+
+    # Ensure close points in original space remain close in embedding
+    group_a_indices = [0, 1, 2]  # Hardcoded index of similar points
+    group_b_indices = [3, 4, 5]  # Hardcoded index of dissimilar points from a
```
How about using the points that were already available and had also differences in signatures - e.g.
```
[2e9, 2e-9, -2e9, -2e-9]
```

yuejiaointel replied:
Hi David,
Thx for the suggestion! Now I picked a few ones like his one in later rows, it works, I'll still run multiple times of ci just to make sure 
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+@pytest.mark.parametrize(
+    "X,n_components,perplexity,expected_shape",
+    [
+        pytest.param(
```
Since there's only one parameterization, maybe these could be defined inside the function body.

yuejiaointel replied:
Hi David,
Thx for the suggestion! Moved parameterized def into function body
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+    dtype,
+):
+    """
+    TSNE test covering multiple functionality and edge cases using parameterization.
```
I think this comment is redundant given the function name.

yuejiaointel replied:
Hi David,
Thx for the suggestion! Removed
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+    # Check for distance b/t two points in group A < distance of this point and any point in group B
+    for i in group_a_indices:
+        for j in group_a_indices:
+            if i != j:
```
This condition should not be needed,

yuejiaointel replied:
Hi David,
Thx for the suggestion! Removed
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+            0.5,
+            (10, 2),
+            False,
+            id="Extremely low perplexity",
```
Considering what this is testing, perhaps it could add a check for no infinites , no NaNs, and no all-zeros in the embeddings in this test too.

yuejiaointel replied:
Hi David,
Thx for the suggestion! Added
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+    X_df = _convert_to_dataframe(X, sycl_queue=queue, target_df=dataframe)
+    tsne = TSNE(n_components=2, perplexity=2.0).fit(X_df)
+    assert "daal4py" in tsne.__module__
+    assert hasattr(tsne, "n_components"), "TSNE missing 'n_components' attribute."
```
This one should already get tested as part of the line that comes next.

yuejiaointel replied:
Hi David,
Thx for the suggestion! removed
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -15,12 +15,256 @@
+    tsne = TSNE(n_components=2, perplexity=2.0).fit(X_df)
+    assert "daal4py" in tsne.__module__
+    assert hasattr(tsne, "n_components"), "TSNE missing 'n_components' attribute."
+    assert tsne.n_components == 2, "TSNE 'n_components' attribute is incorrect."
```
Could also check for 'perplexity' as it was passed to the constructor too.

yuejiaointel replied:
Hi David,
Thx for the suggestion! Added
Best,
Yue

david-cortes-intel commented:
Odd that the CI fails for sklearn1.0 and 1.1. I see that there's many places throughout the code with conditions for sklearn<1.2 though:
https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/daal4py/sklearn/manifold/_t_sne.py
Perhaps something is going wrong there.

yuejiaointel commented:
> Odd that the CI fails for sklearn1.0 and 1.1. I see that there's many places throughout the code with conditions for sklearn<1.2 though: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/daal4py/sklearn/manifold/_t_sne.py Perhaps something is going wrong there.

Hi David!
The tests are failing when checking 0 embedding when data is constant. But actually after I read the implementation I think it make sense. About the version difference, **default initialization in TSNE is changed from random to pca from 1.2** and that is why we see different behavior for different versions. Here is my understanding why we are getting non 0 embedding when we are using self._init = "random":
1. First we are generating a random embedding at first
X_embedded = 1e-4 * random_state.standard_normal(size=(n_samples, n_components))
2. Next the joint probabilities are meaningless since the pairwise distance are 0 but there could still be a value so it will still push the embedding in the gradient decent process and in the end we see some degenerated embedding.
P = _joint_probabilities(distances, self.perplexity, self.verbose)

I actually tested using sklearn and see same behavior here is my simple script and I also see a non zero embedding in the end.
from sklearn.manifold import TSNE
import numpy as np
constant_data = np.full((10, 10), fill_value=1)
tsne = TSNE(n_components=2, init='random',random_state=42, perplexity=5)
embedding = tsne.fit_transform(constant_data)
print(embedding)

I removed the check for 0 embedding because we see zero and non zero embedding for different versions, hope this helps!
Yue

david-cortes-intel commented:
> > Odd that the CI fails for sklearn1.0 and 1.1. I see that there's many places throughout the code with conditions for sklearn<1.2 though: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/daal4py/sklearn/manifold/_t_sne.py Perhaps something is going wrong there.
> 
> Hi David! The tests are failing when checking 0 embedding when data is constant. But actually after I read the implementation I think it make sense. About the version difference, **default initialization in TSNE is changed from random to pca from 1.2** and that is why we see different behavior for different versions. Here is my understanding why we are getting non 0 embedding when we are using self._init = "random":
> 
> 1. First we are generating a random embedding at first
>    X_embedded = 1e-4 * random_state.standard_normal(size=(n_samples, n_components))
> 2. Next the joint probabilities are meaningless since the pairwise distance are 0 but there could still be a value so it will still push the embedding in the gradient decent process and in the end we see some degenerated embedding.
>    P = _joint_probabilities(distances, self.perplexity, self.verbose)
> 
> I actually tested using sklearn and see same behavior here is my simple script and I also see a non zero embedding in the end. from sklearn.manifold import TSNE import numpy as np constant_data = np.full((10, 10), fill_value=1) tsne = TSNE(n_components=2, init='random',random_state=42, perplexity=5) embedding = tsne.fit_transform(constant_data) print(embedding)
> 
> I removed the check for 0 embedding because we see zero and non zero embedding for different versions, hope this helps! Yue

Then please move it into a separate test, using `init="pca"` in the constructor.

Also I don't think the test should look for exact zeros everywhere. For constant data, should test for a constant first dimension, and a zero second dimension.

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -24,3 +33,222 @@ def test_sklearnex_import():
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_tsne_reproducibility(dataframe, queue, dtype):
+    """
+    Test reproducibility
```
Let's better remove obvious or redundant comments. Same for example for error messages in assertions - you will see the line that failed in a log when that happens, and many of those cases will be pretty obvious from the code line, line "isfinite failed".

yuejiaointel replied:
Hi David,
Thx for the advice! I removed all obvious comments
Best,
Yue

yuejiaointel commented:
> > > Odd that the CI fails for sklearn1.0 and 1.1. I see that there's many places throughout the code with conditions for sklearn<1.2 though: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/daal4py/sklearn/manifold/_t_sne.py Perhaps something is going wrong there.
> > 
> > 
> > Hi David! The tests are failing when checking 0 embedding when data is constant. But actually after I read the implementation I think it make sense. About the version difference, **default initialization in TSNE is changed from random to pca from 1.2** and that is why we see different behavior for different versions. Here is my understanding why we are getting non 0 embedding when we are using self._init = "random":
> > 
> > 1. First we are generating a random embedding at first
> >    X_embedded = 1e-4 * random_state.standard_normal(size=(n_samples, n_components))
> > 2. Next the joint probabilities are meaningless since the pairwise distance are 0 but there could still be a value so it will still push the embedding in the gradient decent process and in the end we see some degenerated embedding.
> >    P = _joint_probabilities(distances, self.perplexity, self.verbose)
> > 
> > I actually tested using sklearn and see same behavior here is my simple script and I also see a non zero embedding in the end. from sklearn.manifold import TSNE import numpy as np constant_data = np.full((10, 10), fill_value=1) tsne = TSNE(n_components=2, init='random',random_state=42, perplexity=5) embedding = tsne.fit_transform(constant_data) print(embedding)
> > I removed the check for 0 embedding because we see zero and non zero embedding for different versions, hope this helps! Yue
> 
> Then please move it into a separate test, using `init="pca"` in the constructor.
> 
> Also I don't think the test should look for exact zeros everywhere. For constant data, should test for a constant first dimension, and a zero second dimension.

Hi David,
Thx for the advice, I created a test for constant data and check both random and pca initialization
Best,
Yue

david-cortes-intel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -24,3 +33,213 @@ def test_sklearnex_import():
+
+@pytest.mark.parametrize("init", ["pca", "random"])
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_tsne_constant_data(init, dtype):
```
This one could also test on GPU.

david-cortes-intel reviewed (approved):
LGTM, pending CI results.

ethanglaser commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -24,3 +33,216 @@ def test_sklearnex_import():
+    tsne = TSNE(n_components=2, perplexity=2.0).fit(X_df)
+    assert "daal4py" in tsne.__module__
+    assert tsne.n_components == 2
+    assert tsne.perplexity == 2.0
```
In my opinion the only thing missing from test scope is an exact result check, which I agree would be unrealistic for all tests, but since the import test is small would it be possible to add a specific embedding value check here after transform? Something like we have for pca here: https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/sklearnex/decomposition/tests/test_pca.py#L35-L41

yuejiaointel replied:
Hi Ethan,
That is a good idea, thx for suggestion, I added a test for exact value in import test. Now just waiting to see if CI can pass
Best,
Yue

david-cortes-intel replied:
@ethanglaser TSNE doesn't have a unique global optimum, so adding a test for exact numbers now is likely going to lead to problems later when it gets executed in e.g. newer machines, newer GPUs, etc. But there is a test checking for general goodness of the results at the end, only the criteria it needs to satisfy is not strict.

ethanglaser reviewed (commented):
Great work on this so far! A few small comments from my side but looks like its almost ready for merge.

ethanglaser commented:
/intelci: run

yuejiaointel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -24,3 +33,216 @@ def test_sklearnex_import():
+
+
+@pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues())
+def test_sklearnex_tsne_import(dataframe, queue):
```
Hi Ethan,
Thx for reviewing this! I removed first test and renamed the second import test
Best,
Yue

yuejiaointel commented on sklearnex/manifold/tests/test_tsne.py:
```diff
@@ -24,3 +33,216 @@ def test_sklearnex_import():
     assert "daal4py" in tsne.__module__
+
+
+from sklearnex.manifold import TSNE
```
Hi Ethan,
Thx for the suggestion! I moved the import to inside every test 
Best,
Yue

ethanglaser commented:
/intelci: run
